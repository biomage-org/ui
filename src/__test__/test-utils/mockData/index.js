import generateMockExperimentData from './mockExperimentData';
import generateMockProjects from './mockProjects';
import generateMockSamples from './mockSamples';
import generateMockExperiment from './mockExperiment';

// This file is a work in progress to generate mock data for the endpoints of the test.
// The generated data should be usable as response data to be sent via the mocked API in mockAPI.js
// The module should also expose the generated data that can be imported and used easily in tests
// A ticket has been created to give more thought into this https://biomage.atlassian.net/browse/BIOMAGE-1553

const responseData = {};

const projects = generateMockProjects(2);
responseData.projects = projects;

const projectWithSamples = responseData.projects[0];

responseData.samples = [generateMockSamples(
  projectWithSamples.uuid,
  projectWithSamples.experiments[0],
  3,
)];

const { samples } = responseData.samples[0];
const sampleIds = Object.keys(samples);

projectWithSamples.samples = sampleIds;

responseData.experiments = [generateMockExperiment(
  projectWithSamples.uuid,
  projectWithSamples.experiments[0],
  projectWithSamples.experiments[0],
  projectWithSamples.samples,
)];

responseData.experimentData = generateMockExperimentData(
  projectWithSamples.uuid,
  projectWithSamples.experiments[0],
  projectWithSamples.experiments[0],
  projectWithSamples.samples,
);

export {
  projects,
  samples,
  responseData,
};
