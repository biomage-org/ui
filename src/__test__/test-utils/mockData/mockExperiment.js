import _ from 'lodash';
import fake from '__test__/test-utils/constants';

const projectExperimentTemplate = (projectId, experimentId, experimentName, sampleIds) => (
  {
    lastViewed: fake.MOCK_DATETIME,
    apiVersion: '2.0.0-data-ingest-seurat-rds-automated',
    createdDate: fake.MOCK_DATETIME,
    meta: {
      pipeline: {
        stateMachineArn: 'arn:aws:states:eu-west-1:000000000000:stateMachine:biomage-qc-development-mock997586f1bae0a6d5c5beed65663ec151test',
        executionArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-qc-development-mock997586f1bae0a6d5c5beed65663ec1519de4:422493f3-68a0-4ecb-8b17-0e8cd7a3test',
      },
      gem2s: {
        paramsHash: 'mock57f90e94eeaa82ee6fb7627110828f2etest',
        stateMachineArn: 'arn:aws:states:eu-west-1:000000000000:stateMachine:biomage-gem2s-development-mock997586f1bae0a6d5c5beed65663ec151test',
        executionArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-gem2s-development-mock997586f1bae0a6d5c5beed65663ec1519de4:5f273383-f093-4112-95b0-d76db896test',
      },
      organism: null,
      type: '10x',
    },
    sampleIds,
    description: 'Analysis description',
    experimentId,
    projectId,
    experimentName,
  }
);

const generateMockProjectExperiments = (
  projectId,
  experimentId,
  experimentName,
  sampleIds = [],
  attrs = {},
) => _.merge(
  projectExperimentTemplate(
    projectId,
    experimentId,
    experimentName,
    sampleIds,
  ),
  attrs,
);

export default generateMockProjectExperiments;
