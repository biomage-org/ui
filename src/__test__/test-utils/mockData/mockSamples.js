import fake from '__test__/test-utils/constants';

const mockFileEntry = (fileName, sampleId) => ({
  valid: true,
  path: `${sampleId}/${fileName}`,
  upload: {
    amplifyPromise: null,
    progress: 100,
    status: 'uploaded',
  },
  name: fileName,
  compressed: true,
  lastModified: fake.MOCK_DATETIME,
  fileObject: {
    path: `${sampleId}/${fileName}`,
    size: 100,
  },
  errors: '',
});

const mockSampleTemplate = (projectId, sampleId, idx) => {
  const sampleTemplate = {
    metadata: {},
    createdDate: fake.LAST_MODIFIED,
    species: null,
    name: `Sample ${idx}`,
    files: {},
    lastModified: fake.LAST_MODIFIED,
    type: '10X Chromium',
    complete: false,
    error: false,
    uuid: `${sampleId}`,
    fileNames: [
      'features.tsv.gz',
      'barcodes.tsv.gz',
      'matrix.mtx.gz',
    ],
    projectUuid: `${projectId}`,
  };

  sampleTemplate.fileNames.forEach((fileName) => {
    sampleTemplate.files[fileName] = mockFileEntry(fileName, sampleId);
  });

  return sampleTemplate;
};

const generateMockSamples = (projectId, experimentId, numSamples = 3) => {
  const mockSample = {
    experimentId,
    projectUuid: projectId,
    samples: {},
  };

  for (let idx = 0; idx < numSamples; idx += 1) {
    const newSampleId = `${fake.SAMPLE_ID}-${idx}`;
    const mockSampleEntry = mockSampleTemplate(projectId, newSampleId, idx);

    mockSample.samples[newSampleId] = mockSampleEntry;
  }

  return mockSample;
};

export default generateMockSamples;
