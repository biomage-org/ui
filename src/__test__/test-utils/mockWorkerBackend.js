import * as socketConnectionMocks from 'utils/socketConnection';
import * as seekWorkResponseMocks from 'utils/work/dispatchWorkRequest';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/seekWorkResponse', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/seekWorkResponse');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    seekFromS3: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
seekWorkResponseMocks.seekFromS3.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
