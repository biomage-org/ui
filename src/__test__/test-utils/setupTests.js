import '@testing-library/jest-dom/extend-expect';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import preloadAll from 'jest-next-dynamic';

configure({ adapter: new Adapter() });
jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
