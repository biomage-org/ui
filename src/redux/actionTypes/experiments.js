const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

/**
 * Set active project.
 */
const PROJECTS_SET_ACTIVE = `${EXPERIMENTS}/setActive`;

/**
 * Add project metadata.
 */
const PROJECTS_METADATA_CREATE = `${EXPERIMENTS}/metadataCreate`;

/**
 * Update project metadata.
 */
const PROJECTS_METADATA_UPDATE = `${EXPERIMENTS}/metadataUpdate`;

/**
 * Delete project metadta.
 */
const PROJECTS_METADATA_DELETE = `${EXPERIMENTS}/metadataDelete`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SWITCH,
  PROJECTS_SET_ACTIVE,
  PROJECTS_METADATA_CREATE,
  PROJECTS_METADATA_UPDATE,
  PROJECTS_METADATA_DELETE,
};
