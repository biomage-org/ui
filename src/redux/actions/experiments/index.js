import createExperiment from './createExperiment';
import updateExperiment from './updateExperiment';
import loadExperiments from './loadExperiments';
import loadExperiment from './loadExperiment';
import switchExperiment from './switchExperiment';
import reorderSamples from './reorderSamples';

export {
  createExperiment,
  updateExperiment,
  loadExperiments,
  loadExperiment,
  switchExperiment,
  reorderSamples,
};
