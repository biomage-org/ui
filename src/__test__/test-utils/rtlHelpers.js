import { fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

// dropFilesIntoDropzone takes an input element and simulates droping files into it
// inputElement: this needs to be the actual input element not the dropzone itself, e.g.:
// * <input data-testid='drop-input'/> => screen.getByTestId('drop-input');
// files should be a list of files, e.g.:
// * new File(['content'], 'file.txt', { type: 'text/plain' });
const dropFilesIntoDropzone = async (inputElement, files) => {
  await act(async () => {
    fireEvent.drop(inputElement, {
      target: { files },
    });
  });
};

export {
  // eslint-disable-next-line import/prefer-default-export
  dropFilesIntoDropzone,
};
