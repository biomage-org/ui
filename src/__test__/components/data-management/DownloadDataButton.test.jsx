import React from 'react';
import _ from 'lodash';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
// import thunk from 'redux-thunk';
import { screen, render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import preloadAll from 'jest-next-dynamic';

import { act } from 'react-dom/test-utils';
// import configureMockStore from 'redux-mock-store';

import DownloadDataButton from 'components/data-management/DownloadDataButton';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import downloadFromUrl from 'utils/downloadFromUrl';

// import initialSamplesState from 'redux/reducers/samples/initialState';
// import initialExperimentsState from 'redux/reducers/experiments/initialState';
// import initialExperimentSettingsState from 'redux/reducers/experimentSettings/initialState';
// import { initialExperimentBackendStatus } from 'redux/reducers/backendStatus/initialState';
// import { getBackendStatus } from 'redux/selectors';

import { processingConfig } from '__test__/test-utils/mockData';
import backendStatusData from '__test__/data/backend_status.json';

import fetchWork from 'utils/work/fetchWork';

import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { makeStore } from 'redux/store';
import { loadExperiments } from 'redux/actions/experiments';

import mockAPI, { generateDefaultMockAPIResponses, promiseResponse } from '__test__/test-utils/mockAPI';
import fake from '__test__/test-utils/constants';
import { loadBackendStatus } from 'redux/actions/backendStatus';
// import { loadSamples } from 'redux/actions/samples';
import { loadProcessingSettings } from 'redux/actions/experimentSettings';

// jest.mock('redux/selectors');
jest.mock('utils/pushNotificationMessage');
jest.mock('utils/downloadFromUrl');
jest.mock('utils/work/fetchWork');

// const mockStore = configureMockStore([thunk]);
// const experimentName = 'Experiment 1';
// const experimentDescription = 'Some description';
// const experimentId = 'my-experiment-ðŸ§¬';
// const sample1Uuid = 'sample-1';
// const sample2Uuid = 'sample-2';

// const noDataState = {
//   experiments: {
//     ...initialExperimentsState,
//     name: experimentName,
//     description: experimentDescription,
//     ids: ['experiment-1'],
//     meta: {
//       ...initialExperimentsState.meta,
//       activeExperimentId: experimentId,
//       loading: false,
//     },
//   },
//   experimentSettings: {
//     ...initialExperimentSettingsState,
//   },
//   samples: {
//     ...initialSamplesState,
//   },
// };

// const withDataState = {
//   ...noDataState,
//   experiments: {
//     ...noDataState.experiments,
//     [experimentId]: {
//       ...noDataState.experiments[experimentId],
//       sampleIds: [sample1Uuid, sample2Uuid],
//       metadataKeys: ['metadata-1'],
//     },
//   },
//   experimentSettings: {
//     processing: {
//       cellSizeDistribution: {
//         [sample1Uuid]: {},
//         [sample2Uuid]: {},
//       },
//     },
//   },
// };

const experimentId = `${fake.EXPERIMENT_ID}-0`;

enableFetchMocks();

const mockAPIResponse = _.merge(
  generateDefaultMockAPIResponses(experimentId),
  {
    [`experiments/${experimentId}/processingConfig`]: () => promiseResponse(
      JSON.stringify({ processingConfig }),
    ),
  },
);

describe('DownloadDataButton', () => {
  let storeState;

  beforeAll(async () => {
    await preloadAll();
  });

  beforeEach(() => {
    jest.clearAllMocks(); // Do not mistake with resetAllMocks()!
    fetchMock.resetMocks();

    storeState = makeStore();
  });

  const renderDownloadDataButton = async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          <DownloadDataButton />
        </Provider>,
      );
    });
  };

  const getMenuItems = async () => {
    const menu = await screen.getByText('Download');
    expect(menu).not.toBeDisabled();

    await act(async () => {
      userEvent.click(menu);
    });

    const options = await screen.getAllByRole('menuitem');
    return options;
  };

  it('should render the download data menu', async () => {
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    await act(async () => {
      storeState.dispatch(loadExperiments());
    });
    await act(async () => {
      storeState.dispatch(loadProcessingSettings(experimentId));
    });
    await act(async () => {
      storeState.dispatch(loadBackendStatus(experimentId));
    });

    // await renderDownloadDataButton(withDataState);
    await renderDownloadDataButton();

    const options = await getMenuItems();
    expect(options).toHaveLength(2);
    expect(options[0]).toHaveTextContent('Processed Seurat object (.rds)');
    expect(options[1]).toHaveTextContent('Data Processing settings (.txt)');
    expect(options[0]).toHaveAttribute('aria-disabled', 'false');
    expect(options[1]).toHaveAttribute('aria-disabled', 'false');
  });

  it('Processed Seurat object option is disabled if qc has not succeeded', async () => {
    const backendStatus = _.cloneDeep(backendStatusData);

    backendStatus.pipeline.status = 'FAILED';
    backendStatus.gem2s.status = 'SUCCEEDED';

    const qcFailMockAPIResponse = _.merge(
      mockAPIResponse,
      { [`experiments/${experimentId}/backendStatus`]: () => promiseResponse(JSON.stringify(backendStatus)) },
    );

    fetchMock.mockIf(/.*/, mockAPI(qcFailMockAPIResponse));

    await act(async () => {
      storeState.dispatch(loadExperiments());
    });
    await act(async () => {
      storeState.dispatch(loadProcessingSettings(experimentId));
    });

    await act(async () => {
      storeState.dispatch(loadBackendStatus(experimentId));
    });

    await renderDownloadDataButton();
    const options = await getMenuItems();
    expect(options[0]).toHaveAttribute('aria-disabled', 'true');
    expect(options[1]).toHaveAttribute('aria-disabled', 'false');
  });

  it('Data procesing settings option is disabled if a step misses a sample', async () => {
    const processingConfigMissingSample = _.cloneDeep(processingConfig);
    // Remove settings for one sample
    delete processingConfigMissingSample.cellSizeDistribution[`${fake.SAMPLE_ID}-0`];

    console.log('processingConfigMissingSampleclassifierDebug');
    console.log(processingConfigMissingSample.classifier);

    const stepMissingMockAPIResponse = _.merge(
      mockAPIResponse,
      {
        [`experiments/${experimentId}/processingConfig`]: () => promiseResponse(
          JSON.stringify({ processingConfig: processingConfigMissingSample }),
        ),
      },
    );

    fetchMock.mockIf(/.*/, mockAPI(stepMissingMockAPIResponse));

    await act(async () => {
      storeState.dispatch(loadExperiments());
    });
    await act(async () => {
      storeState.dispatch(loadProcessingSettings(experimentId));
    });

    await act(async () => {
      storeState.dispatch(loadBackendStatus(experimentId));
    });

    await renderDownloadDataButton();
    const options = await getMenuItems();
    expect(options[0]).toHaveAttribute('aria-disabled', 'false');
    expect(options[1]).toHaveAttribute('aria-disabled', 'true');
  });

  it('Downloads data properly', async () => {
    // fetchAPI.mockImplementation(() => Promise.resolve('signedUrl'));
    // getBackendStatus.mockImplementation(() => () => ({
    //   ...initialExperimentBackendStatus,
    //   status: {
    //     pipeline: {
    //       status: 'SUCCEEDED',
    //     },
    //     gem2s: {
    //       status: 'SUCCEEDED',
    //     },
    //   },
    // }));

    fetchWork.mockImplementation((expId, body) => {
      if (body.name === 'GetEmbedding') {
        return Promise.resolve();
      }
    });

    // await renderDownloadDataButton(withDataState);

    await renderDownloadDataButton();

    // Open the Download dropdown
    userEvent.click(screen.getByText(/Download/i));

    const downloadButton = screen.getByText(/Processed Seurat object/i);

    await act(async () => {
      fireEvent.click(downloadButton);
    });

    expect(downloadFromUrl).toHaveBeenCalledTimes(1);
  });

  it('Shows an error if there is an error downloading data', async () => {
    // fetchAPI.mockImplementation(() => Promise.reject(new Error('Something went wrong')));
    // getBackendStatus.mockImplementation(() => () => ({
    //   ...initialExperimentBackendStatus,
    //   status: {
    //     pipeline: {
    //       status: 'SUCCEEDED',
    //     },
    //     gem2s: {
    //       status: 'SUCCEEDED',
    //     },
    //   },
    // }));

    // await renderDownloadDataButton(withDataState);
    await renderDownloadDataButton();

    // Open the Download dropdown
    userEvent.click(screen.getByText(/Download/i));

    const downloadButton = screen.getByText(/Processed Seurat object/i);

    await act(async () => {
      fireEvent.click(downloadButton);
    });

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
  });

  it('Has options disabled if backend status is still loading', async () => {
    // getBackendStatus.mockImplementation(() => () => ({
    //   loading: true,
    //   error: false,
    //   status: null,
    // }));

    // const state = { ...withDataState };
    // await renderDownloadDataButton(state);
    await renderDownloadDataButton();
    const options = await getMenuItems();

    expect(options[0]).toHaveAttribute('aria-disabled', 'true');
    expect(options[1]).toHaveAttribute('aria-disabled', 'true');
  });
});
