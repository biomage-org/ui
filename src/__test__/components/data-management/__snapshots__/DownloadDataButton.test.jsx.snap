// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadDataButton Downloads processed matrix properly 1`] = `
Array [
  Array [
    "testae48e318dab9a1bd0bexperiment-0",
    Object {
      "config": Object {
        "distanceMetric": "cosine",
        "minimumDistance": 0.3,
      },
      "name": "GetEmbedding",
      "type": "umap",
      "useSaved": undefined,
    },
    [Function],
    [Function],
    Object {
      "timeout": 900,
    },
  ],
  Array [
    "testae48e318dab9a1bd0bexperiment-0",
    Object {
      "embeddingETag": "40425699abc7c16fbf3430b3d4ff3f3e",
      "name": "DownloadAnnotSeuratObject",
    },
    [Function],
    [Function],
    Object {
      "timeout": 180,
    },
  ],
]
`;
